openssl-dynamic/src/main/c/io_netty_tcnative_jni_ocsp_Native.c

```diff
@@ -0,0 +1,314 @@
+  }
+
+  if (arg == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no callback object, %i\n", errorCode);
+    return errorCode;
+  }
+
+  JavaVM *jvm = tcn_get_java_vm();
+  if (jvm == NULL) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no JVM, %i\n", errorCode);
+    return errorCode;
+  }
+
+  JNIEnv *env;
+  if (tcn_get_java_env(&env) != JNI_OK) {
+    fprintf(stderr, "io_netty_tcnative_jni_ocsp_callback: no JNIEnv, %i\n", errorCode);
+    return errorCode;
+  }
+
+  if ((*jvm)->AttachCurrentThread(jvm, (void **)&env, NULL) != JNI_OK) {
+    fprintf(stderr, "AttachCurrentThread: %i\n", errorCode);
+    logException(env);
+    return errorCode;
+  }
+
+  jint value = 0;
+  jobject callback = (*env)->NewLocalRef(env, (jweak)arg);
+
+  if ((*env)->IsSameObject(env, callback, NULL) == JNI_FALSE) {
+
+    value = (*env)->CallIntMethod(env, callback, CALLBACK_METHOD, (jlong)ssl);
+
+    if ((*env)->ExceptionCheck(env)) {
+      fprintf(stderr, "CallIntMethod: %i\n", errorCode);
+      logException(env);
+
+      // Use the errorCode in the event of an JVM exception
+      value = errorCode;
+    }
+  }
+
+  (*env)->DeleteLocalRef(env, callback);
+  (*jvm)->DetachCurrentThread(jvm);
+
+  return (int)value;
+}
+
+/**
+ * The client uses -1 for errors
+ */
+static int io_netty_tcnative_jni_ocsp_client_callback(SSL *ssl, void *arg) {
+  return io_netty_tcnative_jni_ocsp_callback(-1, ssl, arg);
+}
+
+/**
+ * The server uses SSL_TLSEXT_ERR_ALERT_FATAL for errors (it's a non-negative number).
+ */
+static int io_netty_tcnative_jni_ocsp_server_callback(SSL *ssl, void *arg) {
+  return io_netty_tcnative_jni_ocsp_callback(SSL_TLSEXT_ERR_ALERT_FATAL, ssl, arg);
+}
+
+jint io_netty_tcnative_jni_ocsp_Native_OnLoad(JNIEnv *env) {
+  jclass clazz = (*env)->FindClass(env, "io/netty/tcnative/jni/ocsp/StatusCallback");
+  if ((*env)->ExceptionCheck(env)) {
+    logException(env);
+    return JNI_ERR;
+  }
+
+  CALLBACK_CLAZZ = (*env)->NewGlobalRef(env, clazz);
+  if ((*env)->ExceptionCheck(env)) {
+    logException(env);
+    return JNI_ERR;
+  }
+
+  CALLBACK_METHOD = (*env)->GetMethodID(env, CALLBACK_CLAZZ, "callback", "(J)I");
+  if ((*env)->ExceptionCheck(env)) {
+    logException(env);
+    return JNI_ERR;
+  }
+
+  return JNI_OK;
+}
+
+void io_netty_tcnative_jni_ocsp_Native_OnUnLoad(JNIEnv *env) {
+  if (CALLBACK_CLAZZ != NULL) {
+    (*env)->DeleteGlobalRef(env, CALLBACK_CLAZZ);
+
+    CALLBACK_CLAZZ = NULL;
+    CALLBACK_METHOD = NULL;
+  }
+}
+
+/**
+ * Installs the OCSP stapling callback.
+ */
+JNIEXPORT jobject JNICALL Java_io_netty_tcnative_jni_ocsp_Native_newCallback
+    (JNIEnv *env, jclass clazz, jlong ctx, jboolean client, jobject callback) {
+
+  long error = -1L;
+  
+  if (ctx == 0L) {
+    fprintf(stderr, "ctx ptr is 0L");
+    return NULL;
+  }
+
+  //
+  // ATTENTION: We need to create a global reference of the callback.
+  //
+  jweak callbackRef = (*env)->NewWeakGlobalRef(env, callback);
```

Comment:
should we control the lifetime more precisely? What if the java lifetime is shorter than the C lifetime (for some reason)? For example have this be cleaned up in `ssl_context_cleanup`.